### PR TITLE
[network] add multi-source support to connectivity manager

### PR DESCRIPTION
**Pull request by phlip9**
To expedite deployment of the new onchain discovery protocol and reduce onchain discovery's dependency on finishing genesis ceremony, this diff allows ConnectivityManager to support multiple discovery sources at once.

Different discovery sources notify the ConnectivityManager of updates to peers' addresses. Currently, there are 3 discovery sources (ordered by decreasing dial priority, i.e., first is highest priority):

1. Onchain discovery protocol
2. Gossip discovery protocol
3. Seed peers from config

In other words, if a we have some addresses discovered via onchain discovery and some seed addresses from our local config, we will try the onchain discovery addresses first and the local seed addresses after.

### CODE REVIEW

File: network/src/connectivity_manager/mod.rs [hunk 1]
```diff
@@ -75,17 +83,35 @@ pub struct ConnectivityManager<TTicker, TBackoff> {
...
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub enum DiscoverySource {
+    Onchain = 0,
+    Gossip = 1,
+    Config = 2,
+}
+
+impl DiscoverySource {
+    const NUM_SOURCES: usize = (Self::Config as u8 + 1) as usize;
```
Comment (bothra90) [inline network/src/connectivity_manager/mod.rs:97]: This leaves room for programmer error where someone might add an element to the list but forget to update this value. A potential "trick" is to define a final entry in the enum list called "Last", although that would mean you'd need to handle that edge case every time you pattern match on a value of DiscoverySource type.

File: network/src/connectivity_manager/mod.rs [hunk 2]
```diff
@@ -438,6 +464,42 @@ fn log_dial_result(peer_id: PeerId, addr: Multiaddr, dial_result: DialResult) {
...
+        self.len() == 0
+    }
+
+    fn update(&mut self, src: DiscoverySource, addrs: Vec<Multiaddr>) {
+        let idx = src as u8 as usize;
+        self.0[idx] = addrs;
+    }
+
+    fn get(&self, mut idx: usize) -> Option<&Multiaddr> {
+        for addrs in &self.0 {
```
Comment (bothra90) [inline network/src/connectivity_manager/mod.rs:492]: might be able to simplify using: `self.0.iter().as_ref().flatten().nth(idx)`

File: network/src/connectivity_manager/mod.rs [hunk 3]
```diff
@@ -75,17 +83,35 @@ pub struct ConnectivityManager<TTicker, TBackoff> {
     event_id: u32,
 }
 
+/// Different sources for peer addresses, ordered by priority (Onchain=highest,
+/// Config=lowest).
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub enum DiscoverySource {
+    Onchain = 0,
```
Comment (bothra90) [inline network/src/connectivity_manager/mod.rs:91]: nit: OnChain?

File: network/src/connectivity_manager/mod.rs (hunk 2 above)
Comment (bothra90) [reply network/src/connectivity_manager/mod.rs:492]: might even be able to do it without `.by_ref()`

File: network/src/connectivity_manager/mod.rs (hunk 1 above)
Comment (phlip9) [reply network/src/connectivity_manager/mod.rs]: I've just realized there's actually a num-variants crate in common/, so i'll use that!

File: network/src/connectivity_manager/mod.rs (hunk 2 above)
Comment (phlip9) [reply network/src/connectivity_manager/mod.rs]: damn, that's pretty big brain. looks like it works too
